larray/core.py

```diff
@@ -6625,29 +6626,29 @@ def percent(self, *axes):
         # I suspect it loose more precision.
         return self * 100 / self.sum(*axes)
 
-    # aggregate method factory
-    def _agg_method(npfunc, nanfunc=None, name=None, commutative=False, by_agg=False):
-        def method(self, *args, **kwargs):
-            keepaxes = kwargs.pop('keepaxes', False)
-            skipna = kwargs.pop('skipna', None)
-            if skipna is None:
-                skipna = nanfunc is not None
-            if skipna and nanfunc is None:
-                raise ValueError("skipna is not available for %s" % name)
-            # func = npfunc
-            func = nanfunc if skipna else npfunc
-            return self._aggregate(func, args, kwargs, by_agg=by_agg,
-                                   keepaxes=keepaxes,
-                                   commutative=commutative)
-        if name is None:
-            name = npfunc.__name__
-        if by_agg:
-            name += "_by"
-        method.__name__ = name
-        return method
-
-    all = _agg_method(np.all, commutative=True)
-    all.__doc__ = """
+    # aggregate method decorator
+    def _decorate_agg_method(npfunc, nanfunc=None, commutative=False, by_agg=False):
+        def decorated(func):
+            def wrapper(self, *args, **kwargs):
+                keepaxes = kwargs.pop('keepaxes', False)
+                skipna = kwargs.pop('skipna', None)
+                if skipna is None:
+                    skipna = nanfunc is not None
+                if skipna and nanfunc is None:
+                    raise ValueError("skipna is not available for {}".format(func.__name__))
+                _npfunc = nanfunc if skipna else npfunc
+                extra_kwargs = {k: kwargs.pop(k) for k in list(kwargs.keys()) if k in _kwarg_agg.keys()}
+                if 'percentile' in func.__name__:
+                    extra_kwargs['q'] = args[0]
```

Comment:
This seems too specific. Wouldn't an some kind of extra argument (eg. fill_extra_kwargs) to _decorate_agg_method be more generic here? which would be given as a dict: {argname: argposition}. For example for percentile, that would be:
{'q': 0}?